src/inmanta/server/services/orchestrationservice.py

```diff
@@ -611,72 +576,85 @@ def safe_get(input: object, key: str, default: object) -> object:
-                date=datetime.datetime.now().astimezone(),
-                total=len(resources),
+                total=len(rid_to_resource),
                 version_info=version_info,
                 undeployable=undeployable_ids,
-                skipped_for_undeployable=skip_list,
+                skipped_for_undeployable=sorted(
+                    self._get_skipped_for_undeployable(list(rid_to_resource.values()), undeployable_ids)
+                ),
                 partial_base=partial_base_version,
+                rids_in_partial_compile=set(rid_to_resource.keys()) if partial_base_version else None,
+                connection=connection,
             )
-            await cm.insert(connection=connection)
         except asyncpg.exceptions.UniqueViolationError:
             raise ServerError("The given version is already defined. Versions should be unique.")
 
-        await data.Resource.insert_many(resource_objects, connection=connection)
-        await cm.update_fields(total=cm.total + len(resources_to_purge), connection=connection)
+        all_resource_version_ids: abc.Set[ResourceVersionIdStr] = set(
+            Id.set_version_in_id(rid, version) for rid in rid_to_resource.keys()
+        )
+        if is_partial_update:
+            # Make mypy happy
+            assert partial_base_version is not None
+            rids_unchanged_resource_sets: abc.Set[
+                ResourceIdStr
+            ] = await data.Resource.copy_resources_from_unchanged_resource_set(
+                environment=env.id,
+                source_version=partial_base_version,
+                destination_version=version,
+                updated_resource_sets=set(sr for sr in resource_sets.values() if sr is not None),
+                deleted_resource_sets=set(removed_resource_sets),
+                connection=connection,
+            )
+            resources_that_moved_resource_sets = rids_unchanged_resource_sets & set(rid_to_resource.keys())
+            if resources_that_moved_resource_sets:
+                raise BadRequest(
+                    f"A partial compile cannot migrate resources {list(resources_that_moved_resource_sets)} to another resource"
+                    " set"
+                )
 
-        for uk in unknowns:
-            await uk.insert(connection=connection)
+            all_resource_version_ids |= {Id.set_version_in_id(rid, version) for rid in rids_unchanged_resource_sets}
 
-        for agent in agents:
-            await self.agentmanager_service.ensure_agent_registered(env, agent, connection=connection)
+        await data.Resource.insert_many(list(rid_to_resource.values()), connection=connection)
+        await cm.recalculate_total(connection=connection)
+
+        await data.UnknownParameter.insert_many(unknowns, connection=connection)
+
+        for res in rid_to_resource.values():
+            await self.agentmanager_service.ensure_agent_registered(env, res.agent, connection=connection)
```

Comment:
I think it would be better to create a set of agents and only call the method for each agent once. The method acquires a session lock and does a query for each invocation, so it's not for free.